### PR TITLE
CMake: add optional LLVM_USE_SHARED_LLVM_LIBRARY option.

### PR DESCRIPTION
**Pull request by LebedevRI**
As discussed in https://github.com/halide/Halide/issues/4284
one important thing is to be able to link to shared versions
of LLVM libraries. This is off-by-default, so i don't believe
this should have any impact on existing users.

The important [existing] caveat to note is that i don't see
where those LLVM libraries-to-be-linked are specified anywhere
in the `halide_config.cmake`, so as long as you are
a. building static halide
and
b. not building halide as a cmake sub-project (via `add_subdirectory()`)
i don't see how the linking can succeed, both either with
old static LLVM libraries, and now with shared LLVM library.

The best i can tell `llvm-config --system-libs` is only needed
when linking to static libraries, to propagate their deps,
so let's not do that for shared LLVM libs.

And the switch to `llvm_config()` is the main thing here,
it internally handles `USE_SHARED`, and does `target_link_libraries()`.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -185,6 +191,7 @@ option(TARGET_OPENGL "Include OpenGL/GLSL target" ON)
 option(TARGET_OPENGLCOMPUTE "Include OpenGLCompute target" ON)
 option(TARGET_D3D12COMPUTE "Include Direct3D 12 Compute target" ON)
 option(HALIDE_SHARED_LIBRARY "Build as a shared library" ON)
+option(HALIDE_SHARED_LLVM_LIBRARY "Use shared versions of LLVM libraries" OFF)
```
Comment (LebedevRI) [inline CMakeLists.txt]: Please bikeshed on the name, are there any better suggestions?

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -758,17 +758,19 @@ else()
...
-# When building a shared library the LLVM libraries will be
-# embedded in the Halide library. When building a static library
-# LLVM is not embedded but CMake knows that when building an executable
-# against the Halide static library that it needs to link LLVM too so
-# PRIVATE scope is the correct choice here.
-target_link_libraries(Halide PRIVATE ${LIBS})
+# When building a shared library the LLVM libraries will either be
+# embedded or linked into the Halide library. When building a static library
+# LLVM is not embedded/linked but CMake knows that when building an executable
+# against the Halide static library that it needs to link LLVM too, IFF the
```
Comment (steven-johnson) [inline src/CMakeLists.txt]: nit: s/the halide/Halide/

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -758,17 +758,19 @@ else()
...
+# halide is being built as a subproject. (which means, statically-built halide
```
Comment (steven-johnson) [inline src/CMakeLists.txt]: Nit: make a proper sentence: "Which means, statically-built Halide doesn't work out-of-installation directory."

File: CMakeLists.txt (hunk 1 above)
Comment (steven-johnson) [reply CMakeLists.txt]: LLVM_USE_SHARED_LLVM_LIBRARY

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -758,17 +758,19 @@ else()
...
+# doesn't work out-of-installation directory)
+# FIXME: migrate to proper cmake config magic.
```
Comment (steven-johnson) [inline src/CMakeLists.txt]: I'd prefer not to land this change until seeing what config magic is actually necessary; it's not clear to me what the FIXME implies.

**Comment by LebedevRI** [comment]
> > The important [existing] caveat to note is that i don't see
> > where those LLVM libraries-to-be-linked are specified anywhere
> > in the `halide_config.cmake`
> 
> Sounds like we need to upgrade the halide_config.cmake contents to specify this, then upgrade halide.cmake to reference whatever we add there as necessary.

Err, not exactly. 

The thing is, `HALIDE_SHARED_LIBRARY` is *already* broken.
As the comment notes, CMake won't link `libLLVM.a` into `libHalide.a`,
nor do you tell the users which LLVM libs should be linked when
linking to `libHalide.a` 'manually'.
`HALIDE_SHARED_LIBRARY` is only not broken when you have access
to the `Halide` CMake target, i.e. when you are building Halide as
a cmake sub-project via `add_subdirectory()` or such.

So `LLVM_USE_SHARED_LLVM_LIBRARY` doesn't change anything here.
This is a separate problem. I believe it **will** be solved further down the road
with later changes, but i'm trying to keep thinks small and simple here.

File: src/CMakeLists.txt (hunk 4 above)
Comment (LebedevRI) [reply src/CMakeLists.txt]: Please note that these issues are not being introduced by this change, they are preexisting.
The FIXME simply records that they are there, and that they likely can be solved.

**Comment by steven-johnson** [comment]
Fair enough. I'll clone this for testing.

**Comment by steven-johnson** [comment]
Thanks for the fix!

**Comment by LebedevRI** [comment]
Thanks!

**Pull request merged**